### PR TITLE
[ML] Multiple items in a single inference request

### PR DESCRIPTION
**Pull request by davidkyle**
Inference requests can be batched by adding more rows to the input tensor. These batch calls are more performant than making multiple calls to `forward()` with a single input when all the inputs are of a similar length. The expected input is now a 2D array of tokens and 2D arrays of supporting arguments, the output is a 3D array. 

This change is purely internal, it does not modify the infer REST API or the interface called by ingest processor but it will be used in planned features.

Requires elastic/ml-cpp#1961

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlParserUtils.java [hunk 1]
```diff
@@ -50,4 +50,58 @@ private MlParserUtils() {}
...
+                    throw new IllegalArgumentException("unexpected token [" + parser.currentToken() + "] for [" + fieldName + "]");
+                }
+
+                if (parser.currentToken() != XContentParser.Token.START_ARRAY) {
+                    throw new IllegalArgumentException("unexpected token [" + parser.currentToken() + "] for [" + fieldName + "]");
+                }
+
+                List<Double> innerInner = new ArrayList<>();
+                while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                    if (parser.currentToken().isValue() == false) {
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlParserUtils.java:85]: I am not sure how you could check the parser if the current token is a double or not. I can only see `isValue`. 

Also, if it weren't a double, would we throw anything other than a parsing exception anyways?

**Review by benwtrent — APPROVED** [review]
My comments have been addressed. @dimitris-athanasiou brought up some good thoughts. 

Once his concerns are addressed, I am good with this PR.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlParserUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlParserUtils.java:85]: We would still throw but I was thinking it gives us a chance for a nicer error message.

**Comment by davidkyle** [comment]
@elasticsearchmachine run elasticsearch-ci/part-2

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlParserUtils.java [hunk 2]
```diff
@@ -50,4 +50,58 @@ private MlParserUtils() {}
...
+
+    /**
+     * Parses a 3 dimensional array of doubles.
+     *
+     * @param fieldName the field name
+     * @param parser the outer parser
+     * @return The 3D array of doubles
+     * @throws IOException
+     */
+    public static double[][][] parseArrayOfArraysOfArraysOfDoubles(String fieldName, XContentParser parser) throws IOException {
```
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlParserUtils.java]: I don't think you're considering the possible names we could have if we follow this convention

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/FillMaskProcessor.java [hunk 3]
```diff
@@ -58,18 +60,19 @@ public void validateInputs(String inputs) {
...
 
-        if (tokenization.getTokens().isEmpty()) {
+        if (tokenization.getTokenizations().isEmpty() ||
+            tokenization.getTokenizations().get(0).getTokens().isEmpty()) {
             return new FillMaskResults(Collections.emptyList());
         }
 
-        int maskTokenIndex = tokenization.getTokens().indexOf(BertTokenizer.MASK_TOKEN);
-        double[] normalizedScores = NlpHelpers.convertToProbabilitiesBySoftMax(pyTorchResult.getInferenceResult()[maskTokenIndex]);
+        int maskTokenIndex = tokenization.getTokenizations().get(0).getTokens().indexOf(BertTokenizer.MASK_TOKEN);
```
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/FillMaskProcessor.java:68]: There is a future use case where we will use this functionality but we don't have it just yet. To make the PR simpler I've limited the processing just the first result of the batch, full batch handling will be added with the use-case. 

I added a TODO comment to explain this

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/NlpTask.java [hunk 4]
```diff
@@ -34,15 +36,61 @@ public NlpTask(NlpConfig config, Vocabulary vocabulary) {
...
+
+                for (; i < tokenization.getLongestSequenceLength(); i++) {
+                    builder.value(padToken);
+                }
+                builder.endArray();
+            }
+            builder.endArray();
+        }
+
+        static void writeNonPaddedIds(String fieldName,
```
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/NlpTask.java]: The function aren't really symmetrical, I've changed it to `writeNonPaddedArguments` which is more accurate.

**Review by dimitris-athanasiou — COMMENTED** [review]
LGTM

**Review by dimitris-athanasiou — APPROVED** [review]

**Pull request merged**